tracer/src/Datadog.Trace/Util/Streams/InitiallyBufferedStream.cs

```diff
@@ -0,0 +1,134 @@
+﻿// <copyright file="InitiallyBufferedStream.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.VendoredMicrosoftCode.System.Buffers;
+
+namespace Datadog.Trace.Util.Streams;
+
+/// <summary>
+/// A stream that only buffers a portion of the initial stream in memory, so that it can be
+/// retrieved again later if needed (for example if deserialization fails)
+/// </summary>
+internal class InitiallyBufferedStream(Stream innerStream) : LeaveOpenDelegatingStream(innerStream)
+{
+    internal const int MaxInitialBufferSize = 128;
+
+    private ArraySegment<byte>? _buffer = null;
+    private int _offset = 0;
+
+    public string? GetBufferedContent()
+    {
+        try
+        {
+            if (_buffer is not null)
+            {
+                return System.Text.Encoding.UTF8.GetString(_buffer.Value.Array!, _buffer.Value.Offset, _buffer.Value.Count);
+            }
+        }
+        catch (Exception)
+        {
+            // Maybe not valid UTF-8, just swallow
+        }
+
+        return null;
+    }
+
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        if (_offset >= _buffer?.Count)
```

Comment:
It's hard for me to parse what `_offset >= _buffer?.Count` will do if _buffer is null.  I would expand the condition

Reply:
I don't know if it's on purpose, but in this code you're only using the buffer for the first read, even if it's not filled entirely. If that's on purpose, then you may just change the condition to `if (_buffer != null)`

Reply:
Fixed and refactored in https://github.com/DataDog/dd-trace-dotnet/pull/5675/commits/0bc52ee67b2f39010667b826bc20e4a3b3b84103